### PR TITLE
Remove the property disabling optimization

### LINKED ISSUE #19677: Turn on optimization for non-windows
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I noticed powershell sets `Optimize=false` on non-windows which might make it slow to execute on those OSs.

 https://github.com/PowerShell/PowerShell/blob/master/PowerShell.Common.props#L189-L194

Can we see if the problem has resolved and turn optimize it back on?

### Expected behavior

```console
Optimize=true
```

### Actual behavior

```console
Optimize=false
```

### Error details

_No response_

### Environment data

```powershell
non-windows
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SeeminglyScience**
# PR Summary

**NOTE:** I'm mainly opening this PR to see if tests fail on any of the platforms we have CI for. If it does, that still tells us where to look.

- Remove the property group that disabled optimization when building for non-Windows platforms
- Set packing size on the `UnixTm` struct to match our native declaration
- Add `SetLastError` to our p/invoke as `libpsl-native!setdate` does set `errno` (without this, I was getting "Unknown Error 203" when running `Set-Date` without sudo. With the change, it displays the expected error)

Resolves #19677

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 1]
```diff
@@ -932,8 +932,9 @@ internal static partial class NativeMethods
                 [LibraryImport(psLib)]
                 internal static partial int WaitPid(int pid, [MarshalAs(UnmanagedType.Bool)] bool nohang);
 
-                // This is a struct tm from <time.h>.
-                [StructLayout(LayoutKind.Sequential)]
+                // This is the struct `private_tm` from setdate.h in libpsl-native.
+                // https://github.com/PowerShell/PowerShell-Native/blob/c5575ceb064e60355b9fee33eabae6c6d2708d14/src/libpsl-native/src/setdate.h#L23
+                [StructLayout(LayoutKind.Sequential, Pack = 4)]
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:937]: Add a comment about Pack = 4

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 2]
```diff
@@ -932,8 +932,9 @@ internal static partial class NativeMethods
                 [LibraryImport(psLib)]
                 internal static partial int WaitPid(int pid, [MarshalAs(UnmanagedType.Bool)] bool nohang);
 
-                // This is a struct tm from <time.h>.
-                [StructLayout(LayoutKind.Sequential)]
+                // This is the struct `private_tm` from setdate.h in libpsl-native.
```
Comment (SeeminglyScience) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:935]:
```suggestion
                // This is the struct `private_tm` from setdate.h in libpsl-native.
                // Packing is set to 4 to match the unmanaged declaration.
```

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs (hunk 1 above)
Comment (SeeminglyScience) [reply src/System.Management.Automation/CoreCLR/CorePsPlatform.cs:938]: Makes sense! Done

**Comment by iSazonov** [comment]
As remainder https://github.com/PowerShell/PowerShell/pull/6881#discussion_r189168467
My suggestion was to use `stime` API to avoid all these issues including ARM.

**Comment by SeeminglyScience** [comment]
> As remainder [#6881 (comment)](https://github.com/PowerShell/PowerShell/pull/6881#discussion_r189168467) My suggestion was to use `stime` API to avoid all these issues including ARM.

Is the issue you're referring to something else that also only occurs when optimization is off? Or is it a different thing we also need to fix in the same area? Asking because for this PR I'm mainly looking at getting optimization back on.

Side note, do you happen to know a good resource going in depth on the layout differences for ARM?

**Review by TravisEz13 — APPROVED** [review]
Should we consider backporting this to 7.3?

**Review by daxian-dbw — APPROVED** [review]
Thanks @SeeminglyScience!

**Comment by iSazonov** [comment]
> > As remainder [#6881 (comment)](https://github.com/PowerShell/PowerShell/pull/6881#discussion_r189168467) My suggestion was to use `stime` API to avoid all these issues including ARM.
> 
> Is the issue you're referring to something else that also only occurs when optimization is off? Or is it a different thing we also need to fix in the same area? Asking because for this PR I'm mainly looking at getting optimization back on.

I pointed the issue to inform you about history. The reason for this problem for many years is that we use a complex structure when we might not. So `stime` uses `time_t` which is usually uint64 which does not require marshalling including ARM. (I haven't checked it since it hasn't been approved before.) 

> 
> Side note, do you happen to know a good resource going in depth on the layout differences for ARM?

I didn't need to dive deep, so I just watched how it was done in the .Net Runtime.
If we continue to use this structure, we need to make sure that this alignment also works on ARM. An example of unpredictability https://stackoverflow.com/questions/5098864/struct-layout-on-wince-gives-unexpected-results That's why I thought of stime.

**Comment by kunalspathak** [comment]
Can you share any benchmark measurements to show how much improvements do we get on linux?

**Comment by daxian-dbw** [comment]
It used to fail on macOS and according to @SeeminglyScience it worked this time. So, we will merge this PR and keep an eye on the release automation test run. If a failure happens there on a specific platform, that will tell us where to look.

**Comment by kunalspathak** [comment]
> It used to fail on macOS and according to @SeeminglyScience it worked this time. So, we will merge this PR and keep an eye on the release automation test run. If a failure happens there on a specific platform, that will tell us where to look.

Did you get a chance to measure the improvements out of it?

**Comment by daxian-dbw** [comment]
No, we didn't do measurement. But it should be [_faster and more efficient_](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/code-generation#optimize) according to the documentation of this compiler option.

**Comment by SeeminglyScience** [comment]
> Did you get a chance to measure the improvements out of it?

Not yet but I do plan to. Busy with some other tasks atm but when I get around to it I'll share in this thread

**Pull request merged**